File: 백준/Silver/15656. N과 M （7）/N과 M （7）.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [N, M] = input[0].split(" ").map(Number);
const nums = input[1]
  .split(" ")
  .map(Number)
  .sort((a, b) => a - b);

const sequence = [];
let result = "";

const dfs = (count) => {
  if (count === M) {
    result += sequence.join(" ") + "\n";
    return;
  }

  for (let i = 0; i < N; i++) {
    sequence.push(nums[i]);
    dfs(count + 1);
    sequence.pop();
  }
};

dfs(0);

console.log(result);

File: 백준/Silver/15656. N과 M （7）/N과 M （7）.py
N, M = map(int, input().split())
nums = list(map(int, input().split()))
stack = []
visited = [0] * (N+1)

nums.sort()


def dfs(depth):
    if depth == M:
        print(*stack)
        return

    for idx, num in enumerate(nums):
        if not visited[idx]:
            stack.append(num)
            dfs(depth+1)
            stack.pop()


dfs(0)
